src/modules/module_21500.c

```diff
@@ -0,0 +1,307 @@
+{
+  u64 *digest = (u64 *) digest_buf;
+
+  pbkdf2_sha1_t *pbkdf2_sha1 = (pbkdf2_sha1_t *) esalt_buf;
+
+  token_t token;
+
+  token.token_cnt  = 3;
+
+  token.signatures_cnt    = 1;
+  token.signatures_buf[0] = SIGNATURE_SOLARWINDS_ORION;
+
+  token.len[0]     = 14;
+  token.attr[0]    = TOKEN_ATTR_FIXED_LENGTH
+                   | TOKEN_ATTR_VERIFY_SIGNATURE;
+
+  token.sep[1]     = '$';
+  token.len_min[1] = 0;
+  token.len_max[1] = 16;
+  token.attr[1]    = TOKEN_ATTR_VERIFY_LENGTH;
+
+  token.sep[2]     = '$';
+  token.len_min[2] = 64;
+  token.len_max[2] = 256;
+  token.attr[2]    = TOKEN_ATTR_VERIFY_LENGTH
+                   | TOKEN_ATTR_VERIFY_BASE64A;
+
+  const int rc_tokenizer = input_tokenizer ((const u8 *) line_buf, line_len, &token);
+
+  if (rc_tokenizer != PARSER_OK) return (rc_tokenizer);
+
+  // iter
+
+  const u32 iter = 1000;
+
+  salt->salt_iter = iter - 1;
+
+  // salt
+
+  char *salt_pos = (char *) token.buf[1];
+  int salt_len = token.len[1];
+
+  if (salt_len > 16) return (PARSER_SALT_LENGTH);
+
+  char custom_salt[17];
+
+  memset (custom_salt, 0, sizeof (custom_salt));
+
+  for (int i = 0; i < salt_len; i++, salt_pos++)
+  {
+    if (*salt_pos >= 'A' && *salt_pos <= 'Z') {
+      custom_salt[i] = *salt_pos + 32;
+    } else {
+      custom_salt[i] = *salt_pos;
+    }
+  }
+
+  if (salt_len < 8)
+  {
+    strncat (custom_salt, SALT_PADDING, 8 - salt_len);
+    salt_len = 8;
+  }
+
+  salt_pos = custom_salt;
+
+  memcpy (pbkdf2_sha1->salt_buf, salt_pos, salt_len);
+
+  salt->salt_len = salt_len;
+
+  salt->salt_buf[0] = pbkdf2_sha1->salt_buf[0];
+  salt->salt_buf[1] = pbkdf2_sha1->salt_buf[1];
+  salt->salt_buf[2] = pbkdf2_sha1->salt_buf[2];
+  salt->salt_buf[3] = pbkdf2_sha1->salt_buf[3];
+  salt->salt_buf[4] = salt->salt_iter;
+
+  // hash
+
+  const u8 *hash_pos = token.buf[2];
+  const int hash_len = token.len[2];
+
+  u8  tmp_buf[256];
+
+  memset (tmp_buf, 0, sizeof (tmp_buf));
+
+  int tmp_len = base64_decode (base64_to_int, hash_pos, hash_len, tmp_buf);
+
+  if (tmp_len < 64) return (PARSER_HASH_LENGTH);
+
+  memcpy (digest, tmp_buf, tmp_len);
+
+  digest[0] = byte_swap_64 (digest[0]);
+  digest[1] = byte_swap_64 (digest[1]);
+  digest[2] = byte_swap_64 (digest[2]);
+  digest[3] = byte_swap_64 (digest[3]);
+  digest[4] = byte_swap_64 (digest[4]);
+  digest[5] = byte_swap_64 (digest[5]);
+  digest[6] = byte_swap_64 (digest[6]);
+  digest[7] = byte_swap_64 (digest[7]);
+
+  return (PARSER_OK);
+}
+
+int module_hash_encode (MAYBE_UNUSED const hashconfig_t *hashconfig, MAYBE_UNUSED const void *digest_buf, MAYBE_UNUSED const salt_t *salt, MAYBE_UNUSED const void *esalt_buf, MAYBE_UNUSED const void *hook_salt_buf, MAYBE_UNUSED const hashinfo_t *hash_info, char *line_buf, MAYBE_UNUSED const int line_size)
+{
+  const u64 *digest = (u64 *) digest_buf;
+
+  const pbkdf2_sha1_t *pbkdf2_sha1 = (const pbkdf2_sha1_t *) esalt_buf;
+
```

Comment:
Typically if you have any strange code to add in the decoder (like you did here):

```
custom_salt[i] = *salt_pos + 32;
```

we expect to do the opposite in the encoder. Why is not here?

Reply:
as described above it's about the lowercase conversion of salt (username). It must not be converted back in the encoder